client/core/core.go

```diff
@@ -458,39 +445,68 @@ func (dc *dexConnection) runMatches(tradeMatches map[order.OrderID]*serverMatche
 		if len(sm.msgMatches) > 0 {
 			err := tracker.negotiate(sm.msgMatches)
 			if err != nil {
-				return fmt.Errorf("negotiate order %v matches failed: %v", oid, err)
+				return updatedAssets, fmt.Errorf("negotiate order %v matches failed: %v", oid, err)
+			}
+
+			// Coins may be returned for canceled orders.
+			tracker.mtx.RLock()
+			if tracker.metaData.Status == order.OrderStatusCanceled {
+				updatedAssets.count(tracker.fromAssetID)
 			}
+			tracker.mtx.RUnlock()
 
 			// Try to tick the trade now, but do not interrupt on error. The
 			// trade will tick again automatically.
-			_, err = tracker.tick()
+			tickUpdatedAssets, err := tracker.tick()
+			updatedAssets.merge(tickUpdatedAssets)
 			if err != nil {
-				return fmt.Errorf("tick of order %v failed: %v", oid, err)
+				return updatedAssets, fmt.Errorf("tick of order %v failed: %v", oid, err)
 			}
 		}
 
-		return nil
+		return updatedAssets, nil
 	}
 
 	// Process the trades concurrently.
 	errChan := make(chan error)
+	assetsUpdatedChan := make(chan assetMap)
 	for _, trade := range tradeMatches {
 		go func(trade *serverMatches) {
-			errChan <- runMatch(trade)
+			assetsUpdated, err := runMatch(trade)
+			assetsUpdatedChan <- assetsUpdated
+			errChan <- err
```

Comment:
You could keep the simpler pattern below with a single `for range` if you instead send and received a `struct{assetMap, error}`.

Reply:
Err.. this was my initial thought but the idea of creating a new struct type within this method didn't sit well with me. Don't see it being used elsewhere, so not sure I should create the type outside the method. I do agree using a struct simplifies things. Will implement.

Reply:
There's no problem with a struct like this just re send the data.  Would must rather do that than create 2 goroutines and have a waitgroup just to received the two pieces of data in parallel.

Reply:
To be clear though, a little struct inside the method is A-OK.